Rationale.md

```diff
@@ -0,0 +1,216 @@
+# Design Rationale
+
+WebAssembly was designed incrementally, and desisions were made with a solid
+rationale which may turn out to be incorrect in the future. Sometimes, decisions
+were made without the luxury of data, and the designers would like to revisit
+certain decisions once real end-to-end usecases are met and Science can be
+performed.
+
+You, the reader, may also be interested in understanding why
+[Abstract Syntax Tree (AST)](AstSemantics.md) semantics are the way they are,
+and digging through github's issues and pull requests becomes difficult.
+
+This rationale document tries to list how decisions were made, and where
+tradeoffs were made for the sake of language ergonomics, portability,
+performance, security, and Getting Things Done.
+
+[Limited local nondeterminism](Nondeterminism.md) is detailed in its own
+document, its rationale is also explained here.
+
+
+## Why AST?
+
+Why not a stack-, register- or SSA-based bytecode?
+* Trees allow a smaller binary encoding: [JSZap][], [Slim Binaries][].
+* [Polyfill prototype][] shows simple and efficient translation to asm.js.
+
+  [JSZap]: https://research.microsoft.com/en-us/projects/jszap/
+  [Slim Binaries]: https://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.108.1711
+  [Polyfill prototype]: https://github.com/WebAssembly/polyfill-prototype-1
+
+
+## Basic Types Only
+
+WebAssembly only represents [a few types](AstSemantics.md#Types).
+* More complex types can be formed from these basic types. It's up to the source
+  language compiler to express its own types in terms of the basic machine
+  types. This allows WebAssembly to present itself as a virtual ISA, and lets
+  compilers target it as they would any other ISA.
+* These types are efficiently executed by all modern architectures.
+* Smaller types (such as `i8` and `i16`) are no more efficient and are only
+  semantically meaningful for memory accesses.
```

Comment:
It might be worth acknowledging that some 8-bit operations are actually faster than their 32-bit counterparts on some x86 chips (Agner). But (a) it's not a *lot* faster, and (b) in C/C++ all arithmetic operands are promoted to int anyway so the only time we see narrower arithmetic happen from C/C++ is when the optimizer proved it was safe, and in many of those cases, it's possible for a wasm backend to prove that it's safe to narrow for the same reason. And (c) these types don't add much over the existing types in terms of functionality, and leaving them out simplifies implementations.